ts/src/encoding/stringEncoder.ts

```diff
@@ -0,0 +1,365 @@
+            count: offsets.length,
+        }),
+        createStream(PhysicalStreamType.LENGTH, encodeVarintInt32Array(dictionaryLengths), {
+            logical: new LogicalStreamType(undefined, undefined, LengthType.DICTIONARY),
+            technique: PhysicalLevelTechnique.VARINT,
+            count: dictionaryLengths.length,
+        }),
+        createStream(PhysicalStreamType.DATA, compressedDictionary, {
+            logical: new LogicalStreamType(DictionaryType.SINGLE),
+        }),
+    );
+}
+
+/**
+ * Encodes a shared dictionary for struct fields.
+ * @param dictionaryStrings - Array of unique strings in the dictionary
+ * @param options - Encoding options
+ * @returns Object containing length and data streams
+ */
+export function encodeSharedDictionary(
+    dictionaryStrings: string[],
+    options: { useFsst?: boolean; dictionaryType?: DictionaryType } = {},
+): {
+    lengthStream: Uint8Array;
+    dataStream: Uint8Array;
+    symbolLengthStream?: Uint8Array;
+    symbolDataStream?: Uint8Array;
+} {
+    const { useFsst = false, dictionaryType = DictionaryType.SHARED } = options;
+
+    const encodedDictionary = encodeStrings(dictionaryStrings);
+    const dictionaryLengths = createStringLengths(dictionaryStrings);
+
+    const lengthStream = createStream(
+        PhysicalStreamType.LENGTH,
+        encodeVarintInt32Array(new Int32Array(dictionaryLengths)),
+        {
+            logical: new LogicalStreamType(undefined, undefined, LengthType.DICTIONARY),
+            technique: PhysicalLevelTechnique.VARINT,
+            count: dictionaryLengths.length,
+        },
+    );
+
+    const dataStream = createStream(PhysicalStreamType.DATA, encodedDictionary, {
+        logical: new LogicalStreamType(dictionaryType),
+        count: encodedDictionary.length,
+    });
+
+    if (useFsst) {
+        const symbolTable = new Uint8Array([99, 97, 116, 100, 111, 103]); // "catdog"
+        const symbolLengths = new Int32Array([3, 3]);
+
+        const symbolLengthStream = createStream(PhysicalStreamType.LENGTH, encodeVarintInt32Array(symbolLengths), {
+            logical: new LogicalStreamType(undefined, undefined, LengthType.SYMBOL),
+            technique: PhysicalLevelTechnique.VARINT,
+            count: symbolLengths.length,
+        });
+
+        const symbolDataStream = createStream(PhysicalStreamType.DATA, symbolTable, {
+            logical: new LogicalStreamType(DictionaryType.FSST),
+            count: symbolTable.length,
+        });
+
+        return { lengthStream, dataStream, symbolLengthStream, symbolDataStream };
+    }
+
+    return { lengthStream, dataStream };
+}
+
+/**
+ * Encodes streams for a struct field.
+ * @param offsetIndices - Indices into the shared dictionary
+ * @param presentValues - Boolean array indicating which values are present
+ * @param isPresent - Whether the field itself is present
+ * @returns Encoded streams for the field
+ */
+export function encodeStructField(
```

Comment:
I can't find the "counter-decode" function for this, in which file is it?
It might be worth a file of its own as I'm not sure how it is related to string...